src/librustc_mir/interpret/place.rs

```diff
@@ -1040,21 +1040,18 @@ where
     ) -> InterpResult<'tcx> {
         let variant_scalar = Scalar::from_u32(variant_index.as_u32()).into();
 
+        if dest.layout.for_variant(self, variant_index).abi.is_uninhabited() {
```

Comment:
I am very unhappy with directly depending on `is_uninhabited`. We don't otherwise do such a check and rely on validation instead.

But it looks like this is otherwise the cleanest solution. Is there a short explanation for why the code below makes no sense for uninhabited variants?

Reply:
> Is there a short explanation for why the code below makes no sense for uninhabited variants?

Uninhabited variants don't get a tag/niche encoding, so there's nothing that could be written, other than undef perhaps.

Reply:
> Uninhabited variants don't get a tag/niche encoding

IOW, layout code uses the *same* check to set up the niche/tag? As long as we can keep that in sync, that seems good to me. Might be best to have two comments (here and in layout) referencing each other to ensure that this is the case.

Reply:
> IOW, layout code uses the *same* check to set up the niche/tag?

Yeah, *exactly* uninhabited variants are ignored when computing the `niche_variants` range or the possible tag values.

I don't think there's anything to keep in sync since there's no possible optimization to make here (uninhabited variants are *by definition* the unrepresented variants), and it might make more sense to keep this code in sync with codegen instead (including e.g. emitting an `unreachable` in codegen instead of just `return`-ing).

I was also suggesting to @oli-obk that this sort of code should be the prime candidate for some sort of "elaborator" that expands MIR into lower-level operations (detailed enough to work for both miri and codegen).

Reply:
> I don't think there's anything to keep in sync since there's no possible optimization to make here (uninhabited variants are by definition the unrepresented variants)

Well layout could switch to some other way of checking inhabitedness.
Is this the relevant call?

https://github.com/rust-lang/rust/blob/2da942f32802c8233a09744024dfbc34431adf65/src/librustc/ty/layout.rs#L1037

There's also this but that takes ZST into account as well:

https://github.com/rust-lang/rust/blob/2da942f32802c8233a09744024dfbc34431adf65/src/librustc/ty/layout.rs#L816-L818

This seems quite odd, looks like at least the "single-variant enum has struct-like layout" only applies of all the other variants are uninhabited *and* ZST?

Reply:
> This seems quite odd, looks like at least the "single-variant enum has struct-like layout" only applies of all the other variants are uninhabited and ZST?

That fixed the bug where you would partially initialize an uninhabited variant and that would leave the bounds of the `enum` and overwrite something else, before getting to the field that actually makes it uninhabited (and therefore would require a panic/infinite loop to attempt to initialize).

> Well layout could switch to some other way of checking inhabitedness.

`for_variant(variant_idx).is_uninhabited()` is *the result* of layout checking inhabitedness for a given variant, *not* the input it uses in doing so.

`.iter().any(|f| f.abi.is_uninhabited())` iterates over the *fields* of the variant (hence the name `f`).

Reply:
> That fixed the bug where you would partially initialize an uninhabited variant and that would leave the bounds of the enum and overwrite something else, before getting to the field that actually makes it uninhabited (and therefore would require a panic/infinite loop to attempt to initialize).

Ah, I keep forgetting that in MIR we can have partial initialization of enums.
(Might be worth having a comment explaining that?)

> for_variant(variant_idx).is_uninhabited() is the result of layout checking inhabitedness for a given variant, not the input it uses in doing so.

Oh I see. All right then.